### PR TITLE
feat(metrics): Improve Prometheus metrics library with custom collectors

### PR DESCRIPTION
**Pull request by lucaslopezf**
### Description

This pull request enhances the existing Prometheus metrics library, extending its capabilities and offering a more tailored and feature-rich metrics collection and reporting framework for Go applications.

#### Key Features

- 📊 **Extended Metrics Types**: Expanded the existing library to support Counter, Gauge, and Histogram types, offering a more comprehensive metrics collection toolkit.
- 🔒 **Type-Safe Operations**: The library ensures type safety, minimizing runtime errors and improving code quality.
- 🏷️ **Dynamic Labeling**: Added the functionality to include custom labels to metrics, allowing for more granular data collection and analysis.
- 🧵 **Enhanced Thread Safety**: Leveraged read-write mutexes to ensure safe concurrent metric updates, preventing data corruption and race conditions.
- 🔄 **Automated Metrics Registration**: With the added feature of automatic metric registration based on types, manual initialization is reduced, making the process more efficient.

### CODE REVIEW

**Review by lawRathod — APPROVED** [review]
LGTM! Great addition, should be really helpful to quickly register metrics using golem. Will try to use it in a project already using golem. 🚀 🚀 🚀

File: pkg/metrics/register.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	t.metrics[name] = MetricDetail{Collector: metric, Handler: handler}
+
+	return nil
+}
+
+func registerCounter(name, help string, labels []string, handler MetricHandler) (prometheus.Collector, error) {
```
Comment (emmanuelm41) [inline pkg/metrics/register.go:43]: Should we check here that the MetricHandler is Counter type as we do with registerHistogram method?

File: pkg/metrics/register.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		if len(labels) == 0 {
+			return prometheus.NewHistogram(prometheus.HistogramOpts{Name: name, Help: help, Buckets: histogramUpdater.Buckets}), nil
+		}
+		return prometheus.NewHistogramVec(prometheus.HistogramOpts{Name: name, Help: help, Buckets: histogramUpdater.Buckets}, labels), nil
+	}
+
+	return nil, fmt.Errorf("invalid handler type for histogram")
+}
+
+func registerGauge(name, help string, labels []string, handler MetricHandler) (prometheus.Collector, error) {
```
Comment (emmanuelm41) [inline pkg/metrics/register.go:66]: Should we check here that the MetricHandler is Gauge type as we do with registerHistogram method?

**Review by emmanuelm41 — COMMENTED** [review]
Just a few comments to review!

File: pkg/metrics/register.go (hunk 1 above)
Comment (lucaslopezf) [reply pkg/metrics/register.go:43]: Yes, we check this here `if _, ok := handler.(*collectors.Counter); !ok {`

File: pkg/metrics/register.go (hunk 2 above)
Comment (lucaslopezf) [reply pkg/metrics/register.go:66]: Yes, we check this here `if _, ok := handler.(*collectors.Gauge); !ok {`

**Pull request merged**